compiler/semstmts.nim

```diff
@@ -859,7 +859,10 @@ proc typeSectionFinalPass(c: PContext, n: PNode) =
       checkConstructedType(s.info, s.typ)
       if s.typ.kind in {tyObject, tyTuple} and not s.typ.n.isNil:
         checkForMetaFields(s.typ.n)
-
+    elif len(a.sons) > 2 and a.sons[1].kind == nkGenericParams and 
```

Comment:
I think we should detect illegal type recursions in the generic instantiations, and not try to do some partial checks for the generic declarations.